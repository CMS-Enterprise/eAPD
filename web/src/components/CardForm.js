import { Alert, Button, Spinner } from '@cmsgov/design-system-core';
import PropTypes from 'prop-types';
import React, { Fragment } from 'react';
import { withRouter } from 'react-router';

const formSubmitNoop = e => e.preventDefault();

const CardForm = ({
  cancelable,
  canSubmit,
  children,
  error,
  footer,
  history,
  legend,
  onSave,
  primaryButtonText: [primaryButtonNormal, primaryButtonWorking],
  sectionName,
  success,
  title,
  working
}) => (
  <div className="card--container">
    <div className="ds-l-container">
      <div className="ds-l-row card">
        <div className="ds-l-col--1 ds-u-margin-left--auto" />
        <div className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6">
          {!!success && <Alert variation="success">{success}</Alert>}
          {!!error && <Alert variation="error">{error}</Alert>}

          <h1 className="ds-h1">
            {sectionName.length > 0 && (
              <span className="ds-h6 ds-u-display--block">
                {sectionName.toUpperCase()}
              </span>
            )}
            {title}
          </h1>
          <form onSubmit={(canSubmit && onSave) || formSubmitNoop}>
            <fieldset className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0">
              {!!legend && <legend className="sr-only">{legend}</legend>}

              {children}

              <div className="ds-u-margin-top--5">
                {onSave && (
                  <Button
                    variation="primary"
                    type="submit"
                    disabled={!canSubmit || working}
                  >
                    {working ? (
                      <Fragment>
                        <Spinner /> {primaryButtonWorking}
                      </Fragment>
                    ) : (
                      primaryButtonNormal
                    )}
                  </Button>
                )}
                {cancelable && (
                  <Button variation="transparent" onClick={history.goBack}>
                    Cancel
                  </Button>
                )}
              </div>
            </fieldset>
          </form>
          {footer && <div className="card--foter">{footer}</div>}
        </div>
        <div className="ds-l-col--1 ds-u-margin-right--auto" />
      </div>
    </div>
  </div>
);

CardForm.propTypes = {
  cancelable: PropTypes.bool,
  canSubmit: PropTypes.bool,
  children: PropTypes.node.isRequired,
  error: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  footer: PropTypes.oneOfType([PropTypes.bool, PropTypes.node]),
  history: PropTypes.object.isRequired,
  legend: PropTypes.string,
  onSave: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  primaryButtonText: PropTypes.arrayOf(PropTypes.string),
  sectionName: PropTypes.string,
  success: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  title: PropTypes.string.isRequired,
  working: PropTypes.bool
};

CardForm.defaultProps = {
  cancelable: true,
  canSubmit: true,
  error: false,
  footer: false,
  legend: '',
  onSave: false,
  primaryButtonText: ['Save changes', 'Working'],
  sectionName: '',
  success: false,
  working: false
};

export default withRouter(CardForm);

export { CardForm as plain };
