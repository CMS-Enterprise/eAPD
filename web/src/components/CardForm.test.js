import { shallow } from 'enzyme';
import React from 'react';
import sinon from 'sinon';

import { plain as CardForm } from './CardForm';

const history = { goBack: sinon.spy() };

describe('card form wrapper', () => {
  beforeEach(() => {
    history.goBack.resetHistory();
  });

  test('renders without a save button if onSave prop is missing', () => {
    expect(
      shallow(
        <CardForm title="test" history={history}>
          hello world
        </CardForm>
      )
    ).toMatchSnapshot();
  });

  test('renders with a save button if onSave prop is provided', () => {
    expect(
      shallow(
        <CardForm title="test" history={history} onSave={sinon.spy()}>
          hello world
        </CardForm>
      )
    ).toMatchSnapshot();
  });

  test('disables the save button if canSubmit is false', () => {
    expect(
      shallow(
        <CardForm
          title="test"
          history={history}
          onSave={sinon.spy()}
          canSubmit={false}
        >
          hello world
        </CardForm>
      )
    ).toMatchSnapshot();
  });

  test('renders a legend if provided', () => {
    expect(
      shallow(
        <CardForm title="test" history={history} legend="of zelda">
          hello world
        </CardForm>
      )
    ).toMatchSnapshot();
  });

  test('renders an error alert if message provided', () => {
    expect(
      shallow(
        <CardForm title="test" history={history} error="oh noes!">
          hello world
        </CardForm>
      )
    ).toMatchSnapshot();
  });

  test('renders a success alert if message provided', () => {
    expect(
      shallow(
        <CardForm title="test" history={history} success="oh yeah!">
          hello world
        </CardForm>
      )
    ).toMatchSnapshot();
  });

  test('does not render a cancel button if form is not cancelable', () => {
    expect(
      shallow(
        <CardForm title="test" history={history} cancelable={false}>
          hello world
        </CardForm>
      )
    ).toMatchSnapshot();
  });

  test('renders a footer if provided', () => {
    expect(
      shallow(
        <CardForm title="test" history={history} footer={<div>Hello</div>}>
          hello world
        </CardForm>
      )
    ).toMatchSnapshot();
  });

  test('renders a spinny-wheel on the save button and disables it if the form is busy', () => {
    expect(
      shallow(
        <CardForm title="test" history={history} onSave={sinon.spy()} working>
          hello world
        </CardForm>
      )
    ).toMatchSnapshot();
  });

  test('calls the onSave prop when the form is submitted', () => {
    const onSave = sinon.spy();
    const component = shallow(
      <CardForm title="test" history={history} onSave={onSave}>
        hello world
      </CardForm>
    );

    component.find('form').simulate('submit');

    expect(onSave.calledOnce).toEqual(true);
  });

  test('calls the onCancel prop when the form is canceled', () => {
    const component = shallow(
      <CardForm title="test" history={history}>
        hello world
      </CardForm>
    );

    component
      .find('Button')
      .findWhere(b => b.prop('variation') === 'transparent')
      .simulate('click');

    expect(history.goBack.calledOnce).toEqual(true);
  });
});
