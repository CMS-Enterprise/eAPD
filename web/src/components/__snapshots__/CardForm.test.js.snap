// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card form wrapper disables the save button if canSubmit is false 1`] = `
<div
  className="card--container"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row card"
    >
      <div
        className="ds-l-col--1 ds-u-margin-left--auto"
      />
      <div
        className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6"
      >
        <h1
          className="ds-h1"
        >
          test
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0"
          >
            hello world
            <div
              className="ds-u-margin-top--5"
            >
              <Button
                disabled={true}
                type="submit"
                variation="primary"
              >
                Save changes
              </Button>
              <Button
                onClick={[Function]}
                type="button"
                variation="transparent"
              >
                Cancel
              </Button>
            </div>
          </fieldset>
        </form>
      </div>
      <div
        className="ds-l-col--1 ds-u-margin-right--auto"
      />
    </div>
  </div>
</div>
`;

exports[`card form wrapper does not render a cancel button if form is not cancelable 1`] = `
<div
  className="card--container"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row card"
    >
      <div
        className="ds-l-col--1 ds-u-margin-left--auto"
      />
      <div
        className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6"
      >
        <h1
          className="ds-h1"
        >
          test
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0"
          >
            hello world
            <div
              className="ds-u-margin-top--5"
            />
          </fieldset>
        </form>
      </div>
      <div
        className="ds-l-col--1 ds-u-margin-right--auto"
      />
    </div>
  </div>
</div>
`;

exports[`card form wrapper renders a footer if provided 1`] = `
<div
  className="card--container"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row card"
    >
      <div
        className="ds-l-col--1 ds-u-margin-left--auto"
      />
      <div
        className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6"
      >
        <h1
          className="ds-h1"
        >
          test
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0"
          >
            hello world
            <div
              className="ds-u-margin-top--5"
            >
              <Button
                onClick={[Function]}
                type="button"
                variation="transparent"
              >
                Cancel
              </Button>
            </div>
          </fieldset>
        </form>
        <div
          className="card--foter"
        >
          <div>
            Hello
          </div>
        </div>
      </div>
      <div
        className="ds-l-col--1 ds-u-margin-right--auto"
      />
    </div>
  </div>
</div>
`;

exports[`card form wrapper renders a legend if provided 1`] = `
<div
  className="card--container"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row card"
    >
      <div
        className="ds-l-col--1 ds-u-margin-left--auto"
      />
      <div
        className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6"
      >
        <h1
          className="ds-h1"
        >
          test
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0"
          >
            <legend
              className="sr-only"
            >
              of zelda
            </legend>
            hello world
            <div
              className="ds-u-margin-top--5"
            >
              <Button
                onClick={[Function]}
                type="button"
                variation="transparent"
              >
                Cancel
              </Button>
            </div>
          </fieldset>
        </form>
      </div>
      <div
        className="ds-l-col--1 ds-u-margin-right--auto"
      />
    </div>
  </div>
</div>
`;

exports[`card form wrapper renders a spinny-wheel on the save button and disables it if the form is busy 1`] = `
<div
  className="card--container"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row card"
    >
      <div
        className="ds-l-col--1 ds-u-margin-left--auto"
      />
      <div
        className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6"
      >
        <h1
          className="ds-h1"
        >
          test
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0"
          >
            hello world
            <div
              className="ds-u-margin-top--5"
            >
              <Button
                disabled={true}
                type="submit"
                variation="primary"
              >
                <Spinner
                  aria-valuetext="Loading"
                  role="progressbar"
                />
                 
                Working
              </Button>
              <Button
                onClick={[Function]}
                type="button"
                variation="transparent"
              >
                Cancel
              </Button>
            </div>
          </fieldset>
        </form>
      </div>
      <div
        className="ds-l-col--1 ds-u-margin-right--auto"
      />
    </div>
  </div>
</div>
`;

exports[`card form wrapper renders a success alert if message provided 1`] = `
<div
  className="card--container"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row card"
    >
      <div
        className="ds-l-col--1 ds-u-margin-left--auto"
      />
      <div
        className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6"
      >
        <Alert
          role="alert"
          variation="success"
        >
          oh yeah!
        </Alert>
        <h1
          className="ds-h1"
        >
          test
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0"
          >
            hello world
            <div
              className="ds-u-margin-top--5"
            >
              <Button
                onClick={[Function]}
                type="button"
                variation="transparent"
              >
                Cancel
              </Button>
            </div>
          </fieldset>
        </form>
      </div>
      <div
        className="ds-l-col--1 ds-u-margin-right--auto"
      />
    </div>
  </div>
</div>
`;

exports[`card form wrapper renders an error alert if message provided 1`] = `
<div
  className="card--container"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row card"
    >
      <div
        className="ds-l-col--1 ds-u-margin-left--auto"
      />
      <div
        className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6"
      >
        <Alert
          role="alert"
          variation="error"
        >
          oh noes!
        </Alert>
        <h1
          className="ds-h1"
        >
          test
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0"
          >
            hello world
            <div
              className="ds-u-margin-top--5"
            >
              <Button
                onClick={[Function]}
                type="button"
                variation="transparent"
              >
                Cancel
              </Button>
            </div>
          </fieldset>
        </form>
      </div>
      <div
        className="ds-l-col--1 ds-u-margin-right--auto"
      />
    </div>
  </div>
</div>
`;

exports[`card form wrapper renders with a save button if onSave prop is provided 1`] = `
<div
  className="card--container"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row card"
    >
      <div
        className="ds-l-col--1 ds-u-margin-left--auto"
      />
      <div
        className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6"
      >
        <h1
          className="ds-h1"
        >
          test
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0"
          >
            hello world
            <div
              className="ds-u-margin-top--5"
            >
              <Button
                disabled={false}
                type="submit"
                variation="primary"
              >
                Save changes
              </Button>
              <Button
                onClick={[Function]}
                type="button"
                variation="transparent"
              >
                Cancel
              </Button>
            </div>
          </fieldset>
        </form>
      </div>
      <div
        className="ds-l-col--1 ds-u-margin-right--auto"
      />
    </div>
  </div>
</div>
`;

exports[`card form wrapper renders without a save button if onSave prop is missing 1`] = `
<div
  className="card--container"
>
  <div
    className="ds-l-container"
  >
    <div
      className="ds-l-row card"
    >
      <div
        className="ds-l-col--1 ds-u-margin-left--auto"
      />
      <div
        className="ds-l-col--12 ds-l-sm-col--10 ds-l-lg-col--6"
      >
        <h1
          className="ds-h1"
        >
          test
        </h1>
        <form
          onSubmit={[Function]}
        >
          <fieldset
            className="ds-u-margin--0 ds-u-padding--0 ds-u-border--0"
          >
            hello world
            <div
              className="ds-u-margin-top--5"
            >
              <Button
                onClick={[Function]}
                type="button"
                variation="transparent"
              >
                Cancel
              </Button>
            </div>
          </fieldset>
        </form>
      </div>
      <div
        className="ds-l-col--1 ds-u-margin-right--auto"
      />
    </div>
  </div>
</div>
`;
